config/config.go

```diff
@@ -120,6 +120,10 @@ var (
 		// Backoff times for retrying a batch of samples on recoverable errors.
 		MinBackoff: model.Duration(30 * time.Millisecond),
 		MaxBackoff: model.Duration(100 * time.Millisecond),
+
+		// Send metric metadata to remote write at the same interval we scrape.
+		SendMetadata:         true,
```

Comment:
I thought a bit about this change, and in theory it makes sense. However, `QueueConfig` is already per endpoint and a sub struct of `RemoteWriteConfig` so I don't really see the value of having them separate. Am I missing something?

In theory, you could have an endpoint just send metadata by tweaking some options (no capacity or shards) but I don't think that is something we want to encourage.

WDYT?

Reply:
> In theory, you could have an endpoint just send metadata by tweaking some options (no capacity or shards) but I don't think that is something we want to encourage.

Yes, but lets say someone want to do this. Imagine suddenly we have 100 big customers wanting this, how we would solve this? Just would be nice to be prepared for this (: If setting shards to 0 would do.. maybe that's ok. 


Also for it's unclear from config standpoint is metadata send as part of normal remote write or out of band? and why out of band? Can I control this? Does any of queue options related to metadata? Shard, backoff, batchsend Deadline, capacity, queue, max samples? All those questions can be a Prometheus maintainer or community burden when people start to ask those questions on our channels. Also it will be impossible to change this config type without introducing compatibility issue, so I would love to take our time here and design something clear. I don't really it should be part of queue config then. Is it really related to queue? And again same boolean comment applies - it's code smell as it does not allow flexibility. ): Not yet sure where is the correct place for this option let's discuss!

Reply:
I also think this needs some more thought. At a minimum I don't think Metadata configs should be embedded inside of the queue config, as the queue config should be around scaling. You could have fields like `MaxMetadataPerSend` in the queue config though, and perhaps that replaces `SendMetadata`?

I don't know what to do about `MetadataSendInterval`, it could probably live in the `RemoteWriteConfig`, or create a `MetadataConfig` off of that? Ideally `MetadataSendInterval` will be deprecated in the future anyway when metadata can be sent alongside the relevant samples.

Reply:
> You could have fields like MaxMetadataPerSend in the queue config though, and perhaps that replaces SendMetadata?

The downside of using something like `MaxMetadataPerSend` is the non-deterministic nature of the data. Is it not sorted at any point so having such an option would send different metadata every time with a potential to never being able to send some data.

> it could probably live in the RemoteWriteConfig, or create a MetadataConfig off of that? Ideally, MetadataSendInterval will be deprecated in the future anyway when metadata can be sent alongside the relevant samples.

I agree! I feel the current structure gives us a way to deprecate without being too invasive in other areas, but I can see how it end up confusing end-users. 

The `QueueConfig` is the only configuration received by the `QueueManager` from where the `MetadataWatcher` spawns. Creating a new `MetadataConfig` implies that a struct which we'll deprecate eventually (and more code to remove) plus a new param to add. I'm happy to proceed with the change, just want to make sure it aligns with what you're both proposing.

Is what I'm hearing, creating a `MetadataConfig` with a `SendInterval` attribute to control the interval and adding a `MaxMetadataPerSend` attribute within `RemoteWriteConfig`. From then, passing `MetadataConfig` to `QueueManager` for using it within `MetadataWatcher` and using something like `if MaxMetadataPerSend > 0` as a way to know whenever metadata is enabled to be sent or not?

Reply:
Yeah, good point on `MaxMetadataPerSend`. Perhaps separate struct is the way to go for now. We already pass things like `walDir` down, so I am not bothered too much about passing another thing down.

Reply:
I have created a separate `MetadataConfig` as suggested, but I couldn't find a way to shy away from using the `boolean` flag, given we want this to be active by default (in that case @bwplotka's suggestion of turn it off if not set is a no-go). I'm happy to implement suggestions as I couldn't come up / find any that would align without a major change to the watcher.

Reply:
I don't see the separate `MetadataConfig` as indicated, did that commit get lost somewhere along the way?

Reply:
This doesn't seem to have been resolved.

Reply:
The last state I recall is that there shouldn't be any configuration here, as the data volumes are so small that it doesn't matter.